File: algorithm/377-combination-sum-iv/377.combination-sum-iv.java
/*
 * @lc app=leetcode id=377 lang=java
 *
 * [377] Combination Sum IV
 */

// @lc code=start
/**
 * comb[target] = sum(comb[target - nums[i]]), where 0 <= i < nums.length, and target >= nums[i]
 * Top down 
 */
class Solution {
    // TC O(nlogn)
    public int combinationSum4(int[] nums, int target) {
        int len = nums.length;
        int []dp = new int[target+1];
        dp[0] = 1;
        Arrays.sort(nums);
        for(int i = 1; i <= target; i++) {
            for(int j = 0; j < nums.length; j++) {
                if(nums[j] > i) {
                    break;
                }
                dp[i] += dp[i-nums[j]];
            }
        }
        return dp[target];
    }
}

// 2025-03-11
class Solution {
    // TC O(k*n)
    public int combinationSum4(int[] nums, int target) {
        int[] dp = new int[target+1]; // how many combination sum at i
        dp[0] = 1; 
        for(int i = 1; i <= target; i++) {
            for( int num : nums) {
                int ni = i-num;
                if(ni >= 0 && ni <= target) {
                    dp[i] += dp[ni];
                }
            }
        }
        return dp[target];
    }
}
// @lc code=end

File: algorithm/377-combination-sum-iv/377.combination-sum-iv.py
#
# @lc app=leetcode id=377 lang=python3
#
# [377] Combination Sum IV
#

# @lc code=start
class Solution:
    def combinationSum4(self, nums: List[int], target: int) -> int:
        def dp_()->int:
            # dp[i], how many way to sum to i with nums
            n = len(nums)
            dp = [0] * (target+1)
            dp[0] = 1
            for i in range(1, target+1):
                for num in nums:
                    if i - num >= 0:
                        dp[i] += dp[i-num]
            return dp[target]
        def dfs()->int:
            n = len(nums)
            self.ans = 0
            nums.sort()
            def helper(nums: List[int], idx: int, sum: int) -> None:
                if nums[idx] + sum == target:
                    self.ans += 1
                elif nums[idx] + sum < target:
                    for i in range(0, n):
                        helper(nums, i, sum+nums[idx])
            for i in range(0, n):
                helper(nums, i, 0)
            return self.ans
        return dp()
# @lc code=end


